File: online_store/domain/models/Catalog.py
from online_store.domain.models.products.ProductInterface import ProductInterface

class Catalog:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(Catalog, cls).__new__(cls)
            cls._instance.items = []
        return cls._instance

    def add_item(self, item: ProductInterface):
        self.items.append(item)

    def search_item(self, sitem: ProductInterface) -> ProductInterface:
        for item in self.items:
            if item == sitem:
                return item
            
        return None
    
    def remove_item(self, item:ProductInterface):
        self.items.remove(item)
        print(f"Removing {item} from catalog")

File: tmpsjava/online_shop/src/domain/models/Catalog.java
package domain.models;

import domain.models.products.ProductInterface;

import java.util.ArrayList;
import java.util.List;

public class Catalog {
    private static Catalog instance = null;
    private final List<ProductInterface> items;

    private Catalog() {
        this.items = new ArrayList<>();
    }

    public static Catalog getInstance() {
        if (instance == null) {
            instance = new Catalog();
        }
        return instance;
    }

    public void add_item(ProductInterface item) {
        this.items.add(item);
    }

    public ProductInterface search_item(ProductInterface new_item) {
        for (ProductInterface item : this.items) {
            if (item.equals(new_item)) {
                return item;
            }
        }
        return null;
    }

    public void remove_item(ProductInterface item) {
        this.items.remove(item);
        System.out.println("Removing " + item + " from catalog");
    }
}
